### ISSUE TITLE
Extend Mailing List Crawler 

### ISSUE THREAD

**Issue by carlosparadis** [sailuh/perceive#70]
## Existing functionality

Currently, our [mailing list crawler](https://github.com/sailuh/perceive/tree/master/Data_Collection/Full_Disclosure) is able to parse most of the content on seclists format. It extracts basically 2 pages format:

 1. The list of threads (e.g. see [this page](http://seclists.org/fulldisclosure/2017/Jan/index.html)). 
 2. The content of a given e-mail reply (e.g. see this [other page](http://seclists.org/fulldisclosure/2017/Jan/3)). 

Textual content, without html tag, is then extracted and saved into a file where the script is executed as a .txt file per e-mail body, and additional information is stored as .csv file (e.g. title, author, timestamp). 

## Known limitations

There are some known limitations: 

In 1) 

 - 1.1 
   - While the existing crawler is able to infer thread and e-mail replies from the list of thread pages using as a simple heuristic "threads that share the same title (when removing expressions such as `Re:`)", it is unable to reconstruct which e-mail is a reply to one another. This limits some of the social network analysis methods that can be used to understand the comunity dynamics. 

In 2)

- 2.1
   - Sometimes, content of the e-mail replies uses different html tags, and the behavior of the crawler is either unknown or undesired. Known tags that occur in e-mail replies are:
       - `<tt>` Currently the crawler ignores this tag. 
      - `<blockquote>`: Behavior is unknown, 

But there may also be other tags unknown to us, and therefore may lead to loss of e-mail content, which is detrimental for later stages of the pipeline for data analysis. 

Additionally, some e-mails contain virus signatures, which should also be parsed out as by themselves hold no semantic meaning. 

- 2.2 E-mail Signatures
   - We are not pre-processing e-mail signatures out. E-mail signatures are problematic for data analysis stages, as the method relies on co-occurrence of words across documents, and this is exactly what e-mail signatures happen to be. Ideally, we would like them to be removed before it is used by the data analysis pipeline.

 - 3 Loss of Raw Data
   - While currently, our interest is in the textual and metadata, there is no guarantee down the road mailing lists such as seclist will not vanish, as it happened before in the same mailing list. If this were to be the case, we currently would be unable to reconstruct the analysis as the true raw data is not stored. 

## Suggested solution 

- (Addresses 3) The current suggested solution is to create a new crawler that download html pages "as-is" into the local machine the script is executed. The extraction of content would be then, rather than done directly from the source website as a crawler, performed on the local machine file, in the sense of a parser, thus enabling new methods to parse the data to be used regardless of the original mailing list. 
- (Addresses 2.2) There exists open source libraries with a collection of heuristics to remove e-mail signatures. One such example is [talon](https://github.com/mailgun/talon). We would like to experiment with such tool to see if it helps to solve 2.2 or if not seek alternative methods. 
- (Addresses 2.1) Creating a parser for the locally stored .html files that identify the various html tags used in the body content, and rank them by frequency. Highest used tags could be sampled to inspect the kind of content that they contain, and then further used as a starting point to create a crawler that specific handles how to parse the content based on our understanding. 
- (Addresses 1) Despite convoluted, the thread page still contains some structure, through indentation, to encode which e-mail thread is a reply to another. Prior students have reported by themselves this heuristic does not suffice, as e-mail threads were combined with different titles. However, a combination of both the current `Re:` removal + same title heuristic and indented titles may yield our desired metadata of the mailing list.  

## Important Note

Attempting to make too many requests to seclists will result in IP Block. Testing the crawler should be done carefully.

**Comment by jgwl** [comment]
I have a spider written for the raw html, but I still need to convert it into Jupyter and supplement it with more notes. 
The main entrypoint is:
`def dl_range(year_start, month_start, year_end, month_end):`

However, my main question is: do you have any preference for file naming conventions? I was going to use a kebab'd four-digit year/abbreviated month (2017-Jan) for the filename, following the mailing list id. Plus:
- *.raw.html for the original source
- *.words.txt for the stripped content
- *.tags.txt for the tag content (addresses 2.1)

Example:
/2017-Jan
-  2017-Jan-0.raw.html
-  2017-Jan-0.words.txt
-  2017-Jan-0.tags.txt

I am also splitting files by month, but was planning on keeping the raw/words/tags files in the same directory. I can split into separate subdirectories if needed. 

Just making sure in case you have any processes downstream that look for specific filenames.

**Comment by carlosparadis** [comment]
> I have a spider written for the raw html, but I still need to convert it into Jupiter and supplement it with more notes.

Hmmm. So, for the crawler, I wouldn't worry so much on the Notebook unless you already have something in mind. Typically this would be something we would prefer as just the script with well-defined args on maybe start and end date, and a clear README.md to use it ([sort of how it is done now for the old crawler](https://github.com/sailuh/perceive/tree/master/Data_Collection/Full_Disclosure), albeit the current interface is not ideal). 

Where I would imagine you would use Notebooks would be afterward, for example when loading the crawled data and trying to apply `talon` and discussing how it is performing etc (i.e. more data analysis). 

Let me know what you think. If you have an idea already, I would be curious to see how you would present it on the Notebook. 

> However, my main question is: do you have any preference for file naming conventions? I was going to use a kebab'd four-digit year/abbreviated month (2017-Jan) for the filename, following the mailing list id. 


We have been using `_` rather than `-`, but there was no rationale behind it nor up to this point a lot of thought was given to the name conventions, so I am glad you are bringing that up. 

To avoid inconsistency with the existing crawler data, I would just recommend sticking with the 2017_Jan_2.< your proposed extension convention>, unless you foresee any problems in using underlines. 

```
*.raw.html for the original source
*.words.txt for the stripped content
*.tags.txt for the tag content (addresses 2.1)
```

This will be a bit tricky. The `.raw.html` is fine for the files, as well as `.tags.txt`, I don't see any major variation of them.  There is a part of the pipeline that groups `e-mail replies into threads`, and `threads into network` csv format for [Gephi](https://gephi.org/). See [this folder](https://github.com/sailuh/perceive/tree/master/Data_Collection/Full_Disclosure/Pre_Processing) that dates back to when the project when was starting, so they have even more rough edges. 

These would be built on top of the .words.txt if you will. But there is another variation point here:

* Recall right now we have on every `.txt` file the **e-mail body**. However, moving forward we may want to also have the title merged into the e-mail. 
* We would also like to account for *with e-mail signature* and *without e-mail signature*. 
* We may also have another variation point on the decision of which tags to use or not (e.g. maybe one will have blockquotes and others will not). 

So the naming scheme needs to account for all these different **setups** when it comes down to the `words.txt`. 

Want to give it a shot on a name convention including all those variants with the possibility to extend them as we add more setups? I don't see the raw.html or the .tags.html being affected, but past the raw data is where any small modification warranties a different collection of files. 

Also, have you considered if it is a good idea to still maintain the metadata once raw.html is processed as a .csv or something else (i.e. author, e-mail, timestamp, etc)?

> I am also splitting files by month, but was planning on keeping the raw/words/tags files in the same directory. I can split into separate subdirectories if needed.

Month folders look fine! When doing the topic modeling that's the granularity we currently use. 

:-) Hopefully, I shared enough food for thought and answered the pending questions. Let me know if I missed anything. 

Would you like to send the code to a Github account so I can give any feedback on that end before the Pull Request? 

Thanks!

**Comment by jgwl** [comment]
> Want to give it a shot on a name convention including all those variants with the possibility to extend them as we add more setups?

How about: 
*.raw.html
*.content.txt : just the stripped out text, no further processing
*.content_nosig.txt (although talon can also identify quoted text, maybe just content_talon?)
*.content_<characteristics>.txt, etc.
*.tags.txt

> Also, have you considered if it is a good idea to still maintain the metadata once raw.html is processed as a .csv or something else (i.e. author, e-mail, timestamp, etc)?

Do you mean keeping metadata in the content file? I plan on extracting metadata both into a sqlite3 db and a csv file, but the content would just include the body, unless you have another preference.

> Would you like to send the code to a Github account so I can give any feedback on that end before the Pull Request?

I've put it in a separate project (https://github.com/jgwl/perceive-fd/blob/master/fd_crawler_raw.py) on my account for now. I wasn't sure how we would handle the entire rewrite of the parser. Should I move the old files into a "v1" subdir, just append my files for now, or just remove the old files?

**Comment by carlosparadis** [comment]
> How about:
*.raw.html
*.content.txt : just the stripped out text, no further processing
*.content_nosig.txt (although talon can also identify quoted text, maybe just content_talon?)
*.content_.txt, etc.
*.tags.txt

I fear on the content piece it will still be unclear whether you are including the body, the body and the title, or just the title, as a stripped out text version of everything wouldn't be used for any portion of the pipeline (as it would include metadata too). Let's try going then with:

*.raw.html
*.title.txt
*.reply.body.txt
*.reply.title_body.txt
*.reply.body_no_signature.txt
*.reply.title_body_no_signature.txt
*.threads.body.txt
*.threads.title_body.txt
*.threads.body_no_signature.txt
*.threads.title_body_no_signature.txt
*.reply.body_tags.txt
*.thread.body_tags.txt
*.threads.edgelist.txt
*.threads.nodes.txt

This is is all the possible combinations of files I can think of for the parsers to output. I thought giving you the exhaustive as it is would be beneficial so you can consider the schema of the database to accommodate them in a meaningful way in how they relate and also to consider how the different parsers may look like. 

Don't worry about the last 2 for now, this is just to keep for reference in a single place. 

> Do you mean keeping metadata in the content file? I plan on extracting metadata both into a sqlite3 db and a csv file, but the content would just include the body, unless you have another preference.

No, currently they are residing in a separate .csv file. Did you get a chance to see them in those yearly folders?

Also, since you were planning on hosting all the data on .sqlite3, and I will likely unable to version it here, how about we just work directly on the final database which I will give you access to on the server? I don't have it in place yet, but I was planning on using MySQL so everyone could reverse engineer using MySql Workbench down the road. Thoughts?

> I've put it in a separate project (https://github.com/jgwl/perceive-fd/blob/master/fd_crawler_raw.py) on my account for now. I wasn't sure how we would handle the entire rewrite of the parser. Should I move the old files into a "v1" subdir, just append my files for now, or just remove the old files?

For now, let's just have your work sit in an entirely separate folder, and once we have a better idea on how the scripts will look like we can decide what to do with it. There is more than just the crawler sitting in that folder as it is since I wanted to reflect the sequence of transformations one can do using the code by nesting them in folders inside the crawler until we had a flat interface referring to everything. 

All the code now resides on Data_Collection in the top folder. Let's go to a new folder on the top level named `Crawlers` and inside it a `Full Disclosure` folder, and another one labeled `Parsers` and inside it a folder called `Full Disclosure` for now. Where `Crawlers/Full Disclosure` would contain for now just your crawler for Full Disclosure html, and `Parsers/Full Disclosure` the other parsers such as the one that extracts the tags, in there.

I will take a look on the code and get back to you after your follow-up on this :-) 

Thanks!

**Comment by carlosparadis** [comment]
I forgot to mention:

## Concerning Code Style

I am trying to move towards using Google Python Styleguide. While Notebooks it is not so much of a deal as the narrative is there (and which becomes the deal in itself), the scripting side usually help keep things saner. 

Usually, the big fish is:

 * https://google.github.io/styleguide/pyguide.html?showone=Naming#Naming
 * https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments
 * https://google.github.io/styleguide/pyguide.html?showone=Imports#Imports

And because this is python:

 * https://google.github.io/styleguide/pyguide.html?showone=Indentation#Indentation

There is also a vim setting if you are a user: https://google.github.io/styleguide/google_python_style.vim

Do you have any objections on us using google's python style guide? It is still work in progress reformatting the old ones on my end, however. 

Also, in general, I think it is ok to sacrifice a bit of performance for clarity (not referring to your code now, but more as a good rule of thumb for the long run). 

This is especially true for us since the large pool of contribution comes from students and our turnover is a semester long. So anything too crazy will slow them down considerably if not make the entire code it's own fortress no man can enter. 

I am just going through this in another project in another lab where the original code owner used all possible fancy methods he could for showcasing, dropped off the face of the earth and hours are spent trying to figure out his code every day to scrape until a new system others can use and understand.

## Concerning handling datetimes and timezones

I had not long ago dreadful experiences with timezone handling in Python. Do you have any preference on using `calendar` and `time`? I was planning on having the codebase staying with time handling using [Arrow](http://arrow.readthedocs.io/en/latest/). Have you used it before? What do you think? 

## Concerning Python Notebooks

I see you are using `requests` to download the data, opposed to something like `scrapy`. I think the way it is, it would be a great candidate for a Python Notebook (I originally thought you would use some framework like scrapy). Here are some examples of how you could go about them I had to make, also using `requests` for data access (although in there we didn't move to Google Python Style yet, planning to):

 * https://github.com/erdl/sensors/blob/master/egauge/Requesting%20raw%20data%20from%20eGauge%20API.ipynb
 * https://github.com/erdl/sensors/blob/master/webctrl/Requesting%20raw%20data%20from%20WebCTRL%20API.ipynb

Of course, feel free to use something else to organize it around, or let me know if you think moving yours to a notebook wouldn't be a good idea.

p.s.: Thanks for accepting the authorization and adding your name! :)

**Comment by jgwl** [comment]
>Concerning Code Style

Glad you have one in mind. I was using numpy style docstrings, but that doesn't work well with markdown. Google should be fine. I really hate how Google aligns indentation to delimiters, but I can work around it 😃 Also, I may cheat from time to time on the 80 char rule...

Also, I always try to pick the simplest/clearest way to do something (for the next person's benefit), so if you see something that looks otherwise, just let me know. 

>Concerning handling datetimes and timezones

My process doesn't really care about timezones; calendar is used to translate "1" into "Jan", and time is just for pausing the thread (so we don't get banned scraping the files). But, using a library will help clarify my date range portion (dl_range). 

BTW, I came across this article about arrow by a competitor, and the bugs pointed out are bad and still not fixed. I am going to try pendulum. 
http://blog.eustace.io/please-stop-using-arrow.html

> Concerning Python Notebooks

I'll work on converting my code to notebooks, and also put it in the new scraper/parser folder structure you asked for.

**Comment by carlosparadis** [comment]
> Glad you have one in mind. I was using numpy style docstrings, but that doesn't work well with markdown. Google should be fine. I really hate how Google aligns indentation to delimiters, but I can work around it 😃 Also, I may cheat from time to time on the 80 char rule... 

Hmm.. I saw there are other style guides, Firefox was one but I know nothing about it. I've just happened to think of Google first :-) Since our code is not styled yet, and I am most familiar with the R version of it if you believe there is another one that may be better feel free to pitch in!
 
I think it would be good if it is something we can also generate documentation out of it, but I will be frank: I never used any of these, and seeing we have already a few different systems in Python, R and even javascript I don't know if it will help or just get in the way. What do you think?

> BTW, I came across this article about arrow by a competitor, and the bugs pointed out are bad and still not fixed. I am going to try pendulum.

Very nice catch, I was unaware. Sure, go for it on Pendulum! And I will make sure to update the notebook I just used Arrow the other day :-) Nothing in PERCEIVE uses arrow, so we are safe as far as I know!

> I'll work on converting my code to notebooks, and also put it in the new scraper/parser folder structure you asked for.

Sounds good! Thank you!

On a side note, I have been reading about the best way to start connecting things around PERCEIVE. The interface as it is is real data. I have aversion to Monolith systems (too many painful experiences already), and the latest thing I saw around as a suggestion was along of `Pipes and Filters`, an Architectural Pattern well known in UNIX and I think Apache NiFi. Would you have any thoughts? This is more for the long run I guess but I thought it wouldn't hurt point it out since you're navigating through the entry point of the entire system.

**Comment by carlosparadis** [comment]
@jgwl Concerning `pendulum`, you may want to check this, as we use `pandas`:

- https://github.com/sdispater/pendulum/issues/131

Which leads to:

- https://github.com/pandas-dev/pandas/issues/15986

`lubridate` in R make this whole thing a breeze, and I am still stuck here trying to make a unix timestamp work:

```
import pendulum
import time

iso_start_datetimez = '2017-09-13 15:19:00'
iso_end_datetimez = '2017-09-13 15:21:00'
tz=-10

start_timestamp = pendulum.parse(iso_start_datetimez,tz=-10)
start_timestamp = time.mktime(start_timestamp.timetuple())
print (start_timestamp)

start_timestamp = pendulum.parse(iso_start_datetimez,tz=-9)
start_timestamp = time.mktime(start_timestamp.timetuple())
print (start_timestamp)
```

Who would have thought a language as popular as python had such frustrations to deal with something as simple as timezones.

**Comment by carlosparadis** [comment]
I opened an issue here: https://github.com/sdispater/pendulum/issues/149

But no clues on the pandas part.

**Comment by jgwl** [comment]
Please take a look at my notebook here: <https://github.com/jgwl/perceive-fd/blob/master/Reply-Parse.ipynb>, and see if that is along the lines of what you were thinking. I'm using the notebook more as a guided tour of the libraries being used, but I will probably continue to create a library alongside these notebooks, to process the bulk of messages.

**Comment by carlosparadis** [comment]
Looks very good. :-) Would you mind opening a Pull Request? That way we can keep the changes and the intermediate commits as we build upon it now that we have a starting point. 

The idea to keep a library in parallel for bulk seems great also.

A few things for this one run:

 - Any reason why you preferred not to have the notebook interpret the `\n`? It would be easier to interpret a bit of what talon is doing when removing content.
 - As far as talon is concerned, you may want to create a simple heuristic to see how it fares against talon brute force if you notice any overall pattern in the signatures.
 - is <a> hyperlink tag? If so, it may also be interesting listing what it is being linked. One of our best examples for the project was found observing a discussion referring other discussions in a few years later in the list. We could play with the content of <a> to identify other such cases, or domains in general referred (I would imagine CVE has a few).

**Comment by jgwl** [comment]
Looking at bugtraq, everything works with just a few changes to the url. I am thinking about renaming the folder (and filenames) to seclists, to represent the mailing list engine instead of the specific list. What is your thought: should I scrap the existing PR and create a new one with the seclists structure?

**Comment by carlosparadis** [comment]
@jgwl You may want to see [this PR](https://github.com/sailuh/topicflow/pull/3#pullrequestreview-75969693) section 7.1.1 only (it's small, don't worry) concerning assumptions on the .csv column names since it now has a (very large) dependency on them. 

Let's maintain the current column names from now onwards, just reformat the way the way the content is parsed as discussed in the previous comment.
